docs/source/cuda/examples.rst

```diff
@@ -3,6 +3,398 @@
+we have a one dimensional object which we'll represent with an array of values. The position 
+of the element in the array is like the position of a small piece of the object, and the value 
+of the element is like the temperature. 
+
+.. literalinclude:: ../../../numba/cuda/tests/doc_examples/test_laplace.py
+   :language: python
+   :caption: from ``test_ex_laplace`` in ``numba/cuda/tests/doc_examples/test_laplace.py``
+   :start-after: ex_laplace.import.begin
+   :end-before: ex_laplace.import.end
+   :dedent: 8
+   :linenos:
+
+
+Some initial setup here. Let's make one point in the center of the object very hot.
+
+.. literalinclude:: ../../../numba/cuda/tests/doc_examples/test_laplace.py
+   :language: python
+   :caption: from ``test_ex_laplace`` in ``numba/cuda/tests/doc_examples/test_laplace.py``
+   :start-after: ex_laplace.allocate.begin
+   :end-before: ex_laplace.allocate.end
+   :dedent: 8
+   :linenos:
+
+The initial state of the problem can be visualized with ``matplotlib``:
+
+.. image:: laplace_initial.svg
+
+In our kernel each thread will be responsible for managing the temperature update for a single element
+in a loop over the desired number of timesteps. The kernel is below. Note the use of cooperative group
+synchronization and the temporary array used to avoid race conditions. See 
+:func:`numba.cuda.cg.this_grid() <numba.cuda.cg.this_grid>` for details.
+
+.. literalinclude:: ../../../numba/cuda/tests/doc_examples/test_laplace.py
+   :language: python
+   :caption: from ``test_ex_laplace`` in ``numba/cuda/tests/doc_examples/test_laplace.py``
+   :start-after: ex_laplace.kernel.begin
+   :end-before: ex_laplace.kernel.end
+   :dedent: 8
+   :linenos:
+
+
+Calling the kernel:
+
+.. literalinclude:: ../../../numba/cuda/tests/doc_examples/test_laplace.py
+   :language: python
+   :caption: from ``test_ex_laplace`` in ``numba/cuda/tests/doc_examples/test_laplace.py``
+   :start-after: ex_laplace.launch.begin
+   :end-before: ex_laplace.launch.end
+   :dedent: 8
+   :linenos:
+
+
+Plotting the final data with a graphing library shows an arc shape highest where
+the object was hot initially and gradually sloping down to zero towards the
+edges where the temperature is fixed at zero. In the limit of infinite time,
+the arc will flatten out completely.
+
+.. image:: laplace_final.svg
+
+.. _cuda_reduction_shared:
+
+Shared Memory Reduction
+=======================
+Numba exposes many CUDA features including shared memory. Shared memory is high speed on chip memory
+that represents a shared address space that all the threads in a block can see. It is local to each
+block and is limited in size, but extremely quick to access, similar to a cache. More can be read 
+about shared memory in the CUDA documentation. As a way of demonstrating shared memory, let's reimplement 
+a famous CUDA solution for summing a vector which works by "folding" the data up using a sucessively
+smaller number of threads.
```

Comment:
I think we can skip explaining shared memory here, and link to the section documenting it:
```suggestion
Numba exposes many CUDA features including :ref:`shared memory
<cuda-shared-memory>`. To demonstrate shared memory, let's reimplement a
famous CUDA solution for summing a vector which works by "folding" the data up
using a sucessively smaller number of threads.

```